pkg/process/monitor/process_monitor.go

```diff
@@ -119,282 +100,239 @@ type ProcessCallback struct {
 //	                                         then the Exit callback will evaluate the same metadata on Exit.
 //	                                         We need to save the metadata here as /proc/pid doesn't exist anymore.
 func GetProcessMonitor() *ProcessMonitor {
-	once.Do(func() {
-		processMonitor = &ProcessMonitor{
-			isInitialized:      false,
-			procEventCallbacks: make(map[ProcessEventType][]*ProcessCallback),
-			runningPids:        make(map[uint32]interface{}),
-		}
-	})
-
 	return processMonitor
 }
 
-func (pm *ProcessMonitor) enqueueCallback(callback *ProcessCallback, pid uint32, metadata interface{}) {
-	if callback.Event == EXEC && callback.Metadata != ANY {
-		switch callback.Metadata {
-		case NAME:
-			pm.runningPids[pid] = metadata
-		}
-	}
-	pm.callbackRunner <- func() { callback.Callback(pid) }
-}
-
-// evalEXECCallback is a best effort and would not return errors, but report them
-func (pm *ProcessMonitor) evalEXECCallback(c *ProcessCallback, pid uint32) {
-	if c.Metadata == ANY {
-		pm.enqueueCallback(c, pid, nil)
-		return
-	}
-
-	proc, err := process.NewProcess(int32(pid))
-	if err != nil {
-		// We receive the Exec event first and /proc could be slow to update
-		end := time.Now().Add(10 * time.Millisecond)
-		for end.After(time.Now()) {
-			proc, err = process.NewProcess(int32(pid))
-			if err == nil {
-				break
-			}
-			time.Sleep(time.Millisecond)
-		}
-	}
-	if err != nil {
-		// short living process can hit here (or later proc.Name() parsing)
-		// as they already exited when we try to find them in /proc
-		// so let's be quiet on the logs as there not much to do here
-		return
-	}
-
-	switch c.Metadata {
-	case NAME:
-		pname, err := proc.Name()
-		if err != nil {
-			log.Debugf("process %d name parsing failed %s", pid, err)
-			return
-		}
-		if c.Regex.MatchString(pname) {
-			pm.enqueueCallback(c, pid, metadataName{Name: pname})
-		}
-	}
-}
+// handleProcessExec is a callback function called on a given pid that represents a new process.
+// we're iterating the relevant callbacks and trigger them.
+func (pm *ProcessMonitor) handleProcessExec(pid int) {
+	pm.processExecCallbacksMutex.RLock()
+	defer pm.processExecCallbacksMutex.RUnlock()
 
-// evalEXITCallback will evaluate the metadata saved by the Exec callback and the callback accordingly
-// please refer to GetProcessMonitor documentation
-func (pm *ProcessMonitor) evalEXITCallback(c *ProcessCallback, pid uint32) {
-	switch c.Metadata {
-	case NAME:
-		metadata, found := pm.runningPids[pid]
-		if !found {
-			// we can hit here if a process started before the Exec callback has been registered
-			// and the process Exit, so we don't find his metadata
-			return
-		}
-		pname := metadata.(metadataName).Name
-		if c.Regex.MatchString(pname) {
-			pm.enqueueCallback(c, pid, metadata)
-		}
-	case ANY:
-		pm.enqueueCallback(c, pid, nil)
+	for callback := range pm.processExecCallbacks {
+		callback := callback
```

Comment:
[nit] I imagine that we're creating a new variable here to keep the reference to the callback alive. However, what do you think of using a different name here to improve readability?